peer-ids/peer-ids.md

```diff
@@ -0,0 +1,139 @@
+# Spec: Peer Ids and Keys
+
+Status: **DRAFT**
+
+## Overview
+
+libp2p uses cryptographic key pairs to sign messages and derive unique
+peer identities (or "peer ids").
+
+This document describes the types of keys supported, how keys are serialized
+for transmission, and how peer ids are generated from the hash of serialized
+public keys.
+
+Although private keys are not transmitted over the wire, the serialization
+format used to store keys on disk is also included as a reference for libp2p
+implementors who would like to import existing libp2p key pairs.
+
+Key encodings and message signing semantics are
+[covered below](#how-keys-are-encoded-and-messages-signed).
+
+## Keys
+
+Our key pairs are wrapped in a [simple protobuf](https://github.com/libp2p/go-libp2p-crypto/blob/master/pb/crypto.proto#L5):
+
+```protobuf
+enum KeyType {
+	RSA = 0;
+	Ed25519 = 1;
+	Secp256k1 = 2;
+	ECDSA = 3;
+}
+
+message PublicKey {
+	required KeyType Type = 1;
+	required bytes Data = 2;
+}
+
+message PrivateKey {
+	required KeyType Type = 1;
+	required bytes Data = 2;
+}
+```
+
+The `PublicKey` and `PrivateKey` messages contain a `Data` field with serialized
+keys, and a `Type` enum that specifies the type of key.
+
+Each key type has its own serialization format within the `Data` field,
+[described below](#how-keys-are-encoded-and-messages-signed).
+
+Note that `PrivateKey` messages are never transmitted over the wire.
+Current libp2p implementations store private keys on disk as a serialized
+`PrivateKey` protobuf message. libp2p implementors who want to load existing
+keys can use the `PrivateKey` message definition to deserialize private key
+files.
+
+#### Where it's used?
+
+Keys are used in two places in libp2p.  The first is for signing messages.  Here are some examples of messages we sign:
+ - IPNS records
+ - PubSub messages (coming soon)
+ - SECIO handshake
+
+The second is for generating peer ids; this is discussed in the section below.
+
+## Peer Ids
+
+Here is the process by which we generate peer id's based on the public/private keypairs described above:
+
+  1. Encode the public key into the protobuf.
+  2. Serialize the protobuf containing the public key into bytes using the [canonical protobuf encoding](https://developers.google.com/protocol-buffers/docs/encoding).
+  3.  If the length of the serialized bytes <= 42, then we compute the "identity" multihash of the serialized bytes.  In other words, no hashing is performed, but the [multihash format is still followed](https://github.com/multiformats/multihash) (byte plus varint plus serialized bytes).  The idea here is that if the serialized byte array is short enough, we can fit it in a multihash verbatim without having to condense it using a hash function.
```

Comment:
This step looks like it could be simplified so that the hashing is always performed - this would simplify peer id management by making them uniform and at the same time disconnect the fingerprint from the actual public key, minimizing the unnecessary spread of key material.

Reply:
Take a look at https://github.com/libp2p/specs/issues/138.

Note: Unfortunately, this isn't something we can "just change" because we have users.